wlcs_anvil/src/ffi_wrappers.rs

```diff
@@ -15,438 +13,156 @@ use wayland_sys::{
+        let surface_id = unsafe { ffi_dispatch!(WAYLAND_CLIENT_HANDLE, wl_proxy_get_id, surface) };
+        if let Some((ref sender, _)) = self.server {
+            let _ = sender.send(WlcsEvent::PositionWindow {
+                client_id,
+                surface_id,
+                location: (x, y).into(),
+            });
         }
     }
 
-    unsafe extern "C" fn create_pointer(ptr: *mut WlcsDisplayServer) -> *mut WlcsPointer {
-        match std::panic::catch_unwind(|| {
-            let me = &mut *container_of!(ptr, DisplayServerHandle, wlcs_display_server);
-            if let Some((ref sender, _)) = me.server {
-                let pointer = Box::into_raw(Box::new(PointerHandle::new(me.next_device_id, sender.clone())));
-                me.next_device_id += 1;
-                &mut (*pointer).wlcs_pointer
-            } else {
-                std::ptr::null_mut()
-            }
-        }) {
-            Ok(ptr) => ptr,
-            Err(err) => {
-                println!(
-                    "panic in wlcs_display_server::create_pointer on ptr: {:p} (type {:?})",
-                    err.as_ref() as *const _,
-                    err.type_id()
-                );
-                std::ptr::null_mut()
-            }
-        }
+    fn create_pointer(&mut self) -> Option<Self::Pointer> {
+        let Some(ref server) = self.server else { return None; };
+        Some(PointerHandle {
+            device_id: DEVICE_ID.fetch_add(1, Ordering::Relaxed),
+            sender: server.0.clone(),
+        })
     }
 
-    unsafe extern "C" fn create_touch(ptr: *mut WlcsDisplayServer) -> *mut WlcsTouch {
-        match std::panic::catch_unwind(|| {
-            let me = &mut *container_of!(ptr, DisplayServerHandle, wlcs_display_server);
-            if let Some((ref sender, _)) = me.server {
-                let pointer = Box::into_raw(Box::new(TouchHandle::new(me.next_device_id, sender.clone())));
-                me.next_device_id += 1;
-                &mut (*pointer).wlcs_touch
-            } else {
-                std::ptr::null_mut()
-            }
-        }) {
-            Ok(ptr) => ptr,
-            Err(err) => {
-                println!(
-                    "panic in wlcs_display_server::create_touch on ptr: {:p} (type {:?})",
-                    err.as_ref() as *const _,
-                    err.type_id()
-                );
-                std::ptr::null_mut()
-            }
-        }
+    fn create_touch(&mut self) -> Option<Self::Touch> {
+        let Some(ref server) = self.server else { return None; };
+        Some(TouchHandle {
+            device_id: DEVICE_ID.fetch_add(1, Ordering::Relaxed),
+            sender: server.0.clone(),
+        })
     }
 
-    unsafe extern "C" fn get_descriptor(_: *const WlcsDisplayServer) -> *const WlcsIntegrationDescriptor {
+    fn get_descriptor(&self) -> *const WlcsIntegrationDescriptor {
```

Comment:
Should this return a &[WlcsIntegrationDescriptor]?

Reply:
Hmm, why not just `WlcsIntegrationDescriptor` actually? What does &[] add?

Reply:
Oh I forgot the array is inside the type.

For returning WlcsIntegrationDescriptor, you'd have to ensure the two internal pointers live as long as the owning instance of the Wlcs trait. I guess just a `&WlcsIntegrationDescriptor` would be fine